usb: add new (experimental) USB device support

jfischer-no opened the pull request:
Add new USB device controller API (UDC API)
Add new USB device controller API and nRF USBD controller driver.
Add new USB device stack implementation using new UDC API

Resolves #19713
Resolves #29132
Resolves #29133
Resolves #29134
Resolves #29135
Resolves #29136
Resolves #30042

_Commits introducing USB host controller (UHC) driver API are here for reasons and will be moved to own PR later._

jfischer-no committed:
drivers: usb: add new USB device controller API (UDC API)

Add new USB device controller API and nRF USBD controller driver.
The new UDC API brings support for multiple instances and asynchronous
transfer model, transfers use net_buf and store methadata in the
user data area.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
include: usb_ch9: add macros to get endpoint attributes

Add macros to get endpoint address, direction, and attributes.
Remove identical endpoint macros from usb_dc header.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: drivers: add simple UDC driver test

Simple test for API rules, allocation, queue, and dequeu
of the endpoint requests. USB device controller should not be
connected to the host as this state is not covered by this test.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

stephanosio commented on include/usb/usb_ch9.h:
```diff
@@ -187,12 +187,27 @@ struct usb_if_descriptor {
 	uint8_t iInterface;
 } __packed;
 
+struct usb_ep_desc_bmattr {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+	uint8_t transfer : 2;
+	uint8_t synch: 2;
+	uint8_t usage: 2;
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+	uint8_t usage : 2;
+	uint8_t synch : 2;
+	uint8_t transfer : 2;
+#endif
+} __packed;
```
We should not use bit fields in the part of code that gets built by multiple toolchains targeting multiple architectures, as the memory layout of C/C++ bit fields is implementation-defined and may not necessarily match the order/size defined here.

jfischer-no replied:
We use it in other subsystems, USB, Bluetooth, 802154. And it has already been discussed. If you still see problems with it then open a general issue. Until then no reason for you to block here.

stephanosio replied:
> We use it in other subsystems, USB, Bluetooth, 802154.

We only have very few instances of bit fields being actually used to read/write structured memory data, and that seems to have been an oversight.

> And it has already been discussed.

Where was it discussed? Can you provide a link to this discussion?

carlescufi replied:
We have discussed this in the past multiple times, just not sure where exactly. Consensus was that, since the use of the `__BYTE_ORDER__ ` is very extended among firmware developers, and is also used in Linux, that we should continue to benefit from the compiler knowing about the bitfields so it can optimize. If this ever turns out to be a real problem (unlikely since this works fine with GCC and Clang, and there aren't many other frontends anymore out there) then we can either add a new macro that deals with it or fall back to masks.

stephanosio replied:
> If this ever turns out to be a real problem (unlikely since this works fine with GCC and Clang, and there aren't many other frontends anymore out there) then we can either add a new macro that deals with it or fall back to masks.

It is true that most compilers follow the order specified in the struct, and there should not be any real issues as long as we make the struct "packed."

However, that does not change the fact that we are at the mercy of the compiler being used, and this may lead to serious portability issues in the future (e.g. this will be a showstopper if we are adding a new toolchain support and that toolchain happens to handle the bit fields in its own implementation-defined way, as permitted by the C standards).

> We have discussed this in the past multiple times, just not sure where exactly. Consensus was that, since the use of the __BYTE_ORDER__ is very extended among firmware developers, and is also used in Linux,

I do not know when that discussion happened or, if any, decision was made; but, if that was a while ago, I would argue that things have changed since then because we now have multiple third-party toolchains (notably Arm Compiler 6 and ARC MWDT) and will likely keep adding new ones in the foreseeable future (e.g. IAR).

One could argue that using non-portable constructs like struct bit fields undermines the efforts made in the toolchain abstraction and portability area.

With that said, the only concern I have at this time is the newly added toolchains such as Arm Compiler 6 and ARC MWDT (cc @tejlmand @evgeniy-paltsev @abrodkin).

carlescufi replied:
@stephanosio this is common in other subsystems as @jfischer-no mentions, including Bluetooth:

https://github.com/zephyrproject-rtos/zephyr/blob/f9db7fe0657b7307e48d42aff8a8030eb4ddecc6/subsys/bluetooth/controller/ll_sw/pdu.h#L375-L380

While you are correct in principle, in practice there is no compiler out there that doesn't respect this (that I know of).

anangl commented on include/drivers/usb/udc.h:
```diff
@@ -0,0 +1,676 @@
+	udc_ep_set_halt_t ep_set_halt;
+	udc_ep_clear_halt_t ep_clear_halt;
+	udc_ep_flush_t ep_flush;
+	udc_ep_buf_free_t ep_free_buf;
+	udc_ep_buf_alloc_t ep_alloc_buf;
```
Shouldn't this be mentioned that these two functions may have no implementation?

jfischer-no replied:
That pars is kind of WiP, I will update that later (TODO)

jfischer-no committed:
usb: add new experimental USB device stack

The device supprt brings support for multiple stack instances,
multiple configuration, asynchronous transfer model, ability to
change most of the properties of a device at runtime and
the composition of configuration and classes at runtime.

The stack requires new UDC driver API and is not compatible
with old driver API (usb_dc_). The classes (functions) of old
(current) USB device stack cannot be used with new ones and must
be ported.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: add shell sample for new USB support

The sample enables new experimental USB device support
and the shell function.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on drivers/usb/udc/Kconfig.nrf:
```diff
@@ -0,0 +1,40 @@
+config UDC_NRF
+	bool "Nordic Semiconductor USB device controller driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_UDC_NRF))
+	depends on HAS_HW_NRF_USBD
```
changed as suggested

jfischer-no commented on drivers/usb/udc/udc_nrf.c:
```diff
@@ -0,0 +1,748 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(udc_nrf, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+K_MSGQ_DEFINE(drv_msgq, sizeof(nrfx_usbd_evt_t), 4, sizeof(uint32_t));
```
fixed

jfischer-no commented on drivers/usb/udc/udc_nrf.c:
```diff
@@ -0,0 +1,748 @@
+LOG_MODULE_REGISTER(udc_nrf, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+K_MSGQ_DEFINE(drv_msgq, sizeof(nrfx_usbd_evt_t), 4, sizeof(uint32_t));
+static K_KERNEL_STACK_DEFINE(drv_stack, 512);
```
fixed

jfischer-no commented on drivers/usb/udc/Kconfig.nrf:
```diff
@@ -0,0 +1,40 @@
+	int "USBD event queue size"
+	range 4 64
+	default 32
+	depends on UDC_NRF
```
changed as suggested

jfischer-no commented on include/drivers/usb/udc.h:
```diff
@@ -0,0 +1,676 @@
+ * @retval -EBUSY request can not be queued
+ * @retval -EPERM controller is not initialized
+ */
+int udc_ep_queue(const struct device *dev, struct net_buf *buf);
```
okay, updated

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,768 @@
+
+bool udc_ep_buf_has_zlp(struct net_buf *buf)
+{
+	struct udc_buf_info *bi;
```
updated

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,768 @@
+
+static bool ep_find_config(const struct device *dev,
+			   struct usb_ep_descriptor *desc,
+			   struct udc_ep_config *cfg)
```
fixed

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,768 @@
+	}
+
+	ret = api->shutdown(dev);
+	data->stat.initialized = false;
```
What one?

jfischer-no commented on drivers/usb/udc/udc_nrf.c:
```diff
@@ -0,0 +1,748 @@
+			ep_cfg_out[i].caps.mps = NRFX_USBD_ISOSIZE / 2;
+		}
+
+		ep_cfg_out[i].addr = i;
```
changed as suggested

jfischer-no committed:
doc: usb: add UDC driver API reference

Add UDC driver API reference.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
soc: expand ifdef by adding new Kconfig option UDC_KINETIS

Expand ifdef by adding new Kconfig option UDC_KINETIS as
preparation for USBFSOTG UDC driver.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc: add USBFSOTG UDC driver for Kinetis SoCs

Add USBFSOTG UDC driver for Kinetis SoCs.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

carlescufi commented:
CC @desowin

desowin commented on include/zephyr/drivers/usb/uhc.h:
```diff
@@ -0,0 +1,477 @@
+ * @brief Get USB host controller capabilities
+ *
+ * Obtain the capabilities of the controller
+ * such as full speed (FS), high speed (HS), and more.
```
Every host controller must support full speed, so I would not mention it here.

desowin commented on subsys/usb/host/usbh_shell.c:
```diff
@@ -0,0 +1,465 @@
+	uint8_t addr;
+	int err;
+
+	addr = strtol(argv[1], NULL, 16);
```
Why base16 for address? Pretty much all documents related to USB describe addresses in decimal.

desowin commented on subsys/usb/host/usbh_shell.c:
```diff
@@ -0,0 +1,465 @@
+	int err;
+
+	addr = strtol(argv[1], NULL, 16);
+	ep = strtol(argv[2], NULL, 16);
```
Same as with address, why base16?

jfischer-no replied:
Because we print/display endpoint address always base16.

desowin reviewed (changes requested):
Reviewed from the USB specification point of view. I did not check any Zephyr specific code (queueing, buffers) as I am not familiar with Zephyr codebase yet.

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1086 @@
+	api->lock(dev);
+
+	ret = ep_check_config(dev, cfg, ep, attributes, *mps, interval);
+	if (ret == true && *mps == 0U) {
```
No, *mps == 0U is the condition to call MPS update

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1086 @@
+
+	switch (ep_attrib_get_transfer(attributes)) {
+	case USB_EP_TYPE_BULK:
+		*mps = MIN(cfg->caps.mps, spec_bulk_mps);
```
How is it not correct when it becomes either 64 or 512?

desowin replied:
Bulk endpoint max packet size depends solely on connection speed and any value provided in `caps.mps` is irrelevant.

The code should be `*mps = spec_bulk_mps;`

desowin replied:
It turns out, I was mistaken! The full-speed bulk max packet size is 8, 16, 32 or 64, that is one of these four explicit values. For high-speed it is 512. Maybe there should be an assert somewhere (preferably build-time).

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1086 @@
+{
+	struct udc_data *data = dev->data;
+	const uint8_t idx = USB_EP_DIR_IS_IN(ep) ?
+			    (ep & BIT_MASK(4)) + 16 :
```
I will not change it.

jfischer-no commented on include/zephyr/drivers/usb/udc.h:
```diff
@@ -0,0 +1,640 @@
+ */
+struct udc_data {
+	/** LUT for endpoint management */
+	struct udc_ep_config *ep_lut[32];
```
I do not.

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1086 @@
+					     struct net_buf *const parent,
+					     const uint8_t ep)
+{
+	size_t size = (ep == USB_CONTROL_EP_OUT) ? 64 : 0;
```
This is backup in case no status comes. For OUT it should be at least setup packet length.

desowin replied:
If there is no status stage then SETUP token ensures device can accept the 8 setup packet bytes. In fact SETUP cancels any incomplete control transfer.

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1086 @@
+			 * complete sequence: s->ack
+			 */
+			LOG_DBG("s->(ack)");
+			next_stage = CTRL_PIPE_STAGE_NO_DATA;
```
Because then one cannot distinguish it while debugging, that is why.

desowin replied:
Why would you have to distinguish it anyway? There are two possible status stages: one with ZLP being sent from device to host, the other one is with ZLP sent from host to device.

The `LOG_DBG("s->(ack)");` seems sufficiently enough debugging information.

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1086 @@
+	return data->stage == CTRL_PIPE_STAGE_STATUS_IN ? true : false;
+}
+
+bool udc_ctrl_stage_is_no_data(const struct device *dev)
```
> There is no bus traffic if SETUP wLength is 0.

What is your statement referring to.

desowin replied:
Consider for example SET CONFIGURATION request to device with address 3.

The SETUP stage consists of single transaction:
```
Host->Device | 2d 03 50                         | SETUP Address 3 Endpoint 0 CRC5: 0x0a
Host->Device | c3 00 09 01 00 00 00 00 00 27 25 | DATA0 (0xc3) Actual Request (8 bytes) CRC16 (0x2527)
Device->Host | d2                               | ACK
```

Because wLength was 0, there is no Data stage and there aren't any transactions on the bus. The state machine directly goes into Status stage.

The Status stage can contain multiple NAKed IN transactions:
```
Host->Device | 69 03 50 | IN Address 3 Endpoint 0 CRC5: 0x0a
Device->Host | 5a       | NAK
```

The Status stage ends with acknowledged IN transaction:
```
Host->Device | 69 03 50 | IN Address 3 Endpoint 0 CRC5: 0x0a
Device->Host | 4b 00 00 | DATA1 (ZLP) CRC: 0x0000
Host->Device | d2       | ACK
```

jfischer-no commented on drivers/usb/udc/udc_common.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * @return pointer to driver's private data
+ */
+static inline void *udc_get_privat(const struct device *dev)
```
Actually it should be `udc_get_private()`

jfischer-no commented on drivers/usb/udc/udc_common.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * @return 0 on success, all other values should be treated as error.
+ */
+int udc_ctrl_submit_s_ack(const struct device *dev);
```
It does not make it more complicated, but more traceable and helps especially during debugging and driver development.

desowin replied:
I disagree. The less states a state machine has the easier it is to debug. This is completely artificial state and I do not understand how it makes things more traceable.

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1086 @@
+{
+	struct udc_device_caps caps = udc_caps(dev);
+	const uint16_t spec_iso_mps = caps.hs ? 1024 : 1023;
+	const uint16_t spec_int_mps = caps.hs ? 1024 : 64;
```
No low speed, I will remove caps.fs.

jfischer-no commented on drivers/usb/udc/udc_common.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * @return true if stage is Status no-data
+ */
+bool udc_ctrl_stage_is_no_data(const struct device *dev);
```
No

jfischer-no commented on include/zephyr/usb/usbd.h:
```diff
@@ -0,0 +1,683 @@
+	/** slist node struct */
+	sys_snode_t node;
+	/** Optional descriptor index, required for string descriptors */
+	uint32_t idx;
```
Yes, stuct is not packed, it does not change anything.

desowin replied:
Code is for humans not compiler. The type should reflect actual data size.

jfischer-no commented on include/zephyr/usb/usbd.h:
```diff
@@ -0,0 +1,683 @@
+	/** Protocol state of the USB device stack */
+	enum usbd_ch9_state state;
+	/** Halted endpoints bitmap */
+	uint32_t ep_halt;
```
No, this is uniform always single uint32_t for all endpoint bitmap variables. I also had it with uint16_t in the past but it did not convince me.

desowin replied:
If you don't want to separate it, then please create macro for index calculation based on endpoint direction and number.

jfischer-no commented on include/zephyr/usb/usbd.h:
```diff
@@ -0,0 +1,683 @@
+ *
+ * @return 0 on success, other values on fail.
+ */
+int usbd_device_set_class(struct usbd_contex *const uds_ctx,
```
Maybe, i think i had it that way in the beginning.

jfischer-no commented on subsys/usb/device_next/class/loopback.c:
```diff
@@ -0,0 +1,296 @@
+		.bInterval = 0x00,				\
+	},							\
+								\
+	/* Interface descriptor 0 */				\
```
Yes

jfischer-no commented on subsys/usb/device_next/usbd_ch9.c:
```diff
@@ -0,0 +1,847 @@
+		}
+	} else {
+		if (!setup->wLength) {
+			LOG_ERR("device-to-host with wLength zero");
```
However, I would like to treat this abnormal behavior as an error.

desowin replied:
But it is not an error! It is perfectly fine. Treating it as error is going to waste a lot of someone's time in the future when troubleshooting Zephyr based device with some host.

desowin commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1086 @@
+		*mps = MIN(cfg->caps.mps, spec_iso_mps);
+		break;
+	case USB_EP_TYPE_CONTROL:
+		*mps = 64U;
```
Then remove the `fs` member from caps as it makes no sense to have it.

jfischer-no commented on subsys/usb/device_next/usbd_class.h:
```diff
@@ -0,0 +1,96 @@
+ *
+ * Calculate the length of the class instance descriptor.
+ * The descriptor must be terminated by a termination descriptor
+ * (bLength = 0 and bDescriptorType = 0).
```
Yes, but that refers to our internal "terminology", `struct usb_desc_header term_desc = {bLength = 0, bDescriptorType = 0,}`

desowin replied:
I would rename the `term_desc` to better reflect its purpose (sentinel). Conflicting project and domain terminology is unnecessarily making things harder to understand.

jfischer-no commented on subsys/usb/device_next/usbd_core.c:
```diff
@@ -0,0 +1,220 @@
+{
+	int ret = 0;
+
+	switch (event->type) {
```
No support planned for the application (user) callbacks, this is to be handled by an event management system.

jfischer-no commented on include/zephyr/usb/usbd.h:
```diff
@@ -0,0 +1,683 @@
+ *
+ * @return 0 on success, other values on fail.
+ */
+int usbd_device_set_vid(struct usbd_contex *const uds_ctx,
```
Yes, but it es valid to disable, change PID and enable device support.

desowin replied:
> Yes, but it es valid to disable, change PID and enable device support.

How does it address the comment? VID and PID form a tuple and should be set only as a tuple.

desowin reviewed (changes requested):
The code is not compliant with Universal Serial Bus Specification Revision 2.0 9.3.1 bmRequestType. See https://github.com/zephyrproject-rtos/zephyr/pull/38679#discussion_r965615586

desowin reviewed (changes requested):
The comment about [Universal Serial Bus Specification Revision 2.0 9.3.1 bmRequestType](https://github.com/zephyrproject-rtos/zephyr/pull/38679/files#r965615586) is still not addressed.

pdgendt commented on drivers/usb/udc/Kconfig.kinetis:
```diff
@@ -0,0 +1,19 @@
+config UDC_KINETIS
+	bool "Kinetis USB device controller driver"
+	imply UDC_WORKQUEUE
+	default $(dt_compat_enabled,$(DT_COMPAT_UDC_KINETIS))
```
This should use the `DT_HAS_..._ENABLED` helpers.

```suggestion
	default y
	depends on DT_HAS_NXP_KINETIS_USBD_ENABLED
```

jfischer-no replied:
Thanks, I was lazy and forgot.

pdgendt commented on drivers/usb/udc/Kconfig.nrf:
```diff
@@ -0,0 +1,29 @@
+
+config UDC_NRF
+	bool "Nordic Semiconductor USB device controller driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_UDC_NRF))
```
Same here

```suggestion
	default y
	depends on DT_HAS_NORDIC_NRF_USBD_ENABLED
```

pdgendt commented on drivers/usb/udc/Kconfig:
```diff
@@ -0,0 +1,54 @@
+config UDC_WORKQUEUE
+	bool "Use a dedicate work queue for UDC drivers"
+	help
+	  This option provides a dedicated work queue that is used for
+	  Without this the system work queue is used for all USB
+	  offloaded transfers.
```
Weird/incomplete description.

pdgendt commented on drivers/usb/udc/Kconfig.kinetis:
```diff
@@ -0,0 +1,18 @@
+	int "Number or blocks in event slab"
+	depends on UDC_KINETIS
+	range 4 16
+	help
+	default 4
+	  Number of blocks in slab for internal endpoint events.
```
Fix ordering.

pdgendt commented on drivers/usb/udc/Kconfig.nrf:
```diff
@@ -0,0 +1,28 @@
+	select NRFX_USBD
+	select NRFX_POWER
+	help
+	  nRF USB device controller driver
```
Nit: descriptions should end with a `.`.

pdgendt commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1112 @@
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_DATA_OUT ? true : false;
```
Why still use the ternary operator?

```suggestion
	return data->stage == CTRL_PIPE_STAGE_DATA_OUT;
```

jfischer-no replied:
Because the type `data->stage == CTRL_PIPE_STAGE_DATA_OUT` is an int not bool.

pdgendt commented on drivers/usb/udc/udc_common.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * This function can be used by the driver to enable control IN/OUT endpoint.
+ *
+ * @param[in] dev    Pointer to device struct of the driver instance
```
Nit: alignment

pdgendt commented on subsys/usb/device_next/usbd_class_api.h:
```diff
@@ -0,0 +1,182 @@
+		return api->control(node, setup, buf);
+	}
+
+	errno = -ENOTSUP;
+	return 0;
```
Did you mean to `return -ENOTSUP;` ?

jfischer-no replied:
No.

pdgendt commented on subsys/usb/device_next/usbd_device.c:
```diff
@@ -0,0 +1,289 @@
+
+	usbd_device_lock(uds_ctx);
+
+	if (!caps.rwup) {
+		LOG_ERR("Remote wakeup feature not supported");
+		return -ENOTSUP;
+	}
+
+	if (!uds_ctx->status.rwup || !usbd_is_suspended(uds_ctx)) {
+		LOG_ERR("Remote wakeup feature not enabled or not suspended");
+		return -EACCES;
+	}
```
Early returns without `usbd_device_unlock`.

pdgendt commented on subsys/usb/device_next/usbd_device.c:
```diff
@@ -0,0 +1,289 @@
+
+	if (!usbd_is_enabled(uds_ctx)) {
+		LOG_WRN("USB device support is already disabled");
+		return 0;
```
```suggestion
		return -EALREADY;
```

pdgendt commented on subsys/usb/device_next/usbd_device.c:
```diff
@@ -0,0 +1,289 @@
+
+	usbd_device_unlock(uds_ctx);
+
+	return 0;
```
```suggestion
	return ret;
```

pdgendt reviewed (commented):
Not an in-depth review, but some remarks while skimming through.

jfischer-no committed:
usb: device_next: add experimental CDC ACM implementation

Add experimental CDC ACM implementation for new USB device stack.
It currently implements only UART IRQ API support and is WIP.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dts: bindings: add fifo size properties to CDC ACM UART bindings

Add RX and TX fifo size properties to CDC ACM UART bindings.
This allows per instance settings in contrast to Kconfig
USB_CDC_ACM_RINGBUF_SIZE option. New properties takes the default
size value of USB_CDC_ACM_RINGBUF_SIZE which can be removed
subsequently.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
include: usb_cdc: add notification codes and UART state bitmap values

Add PSTN subclass notification codes and UART state bitmap values.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: cdc_acm: enable optional new USB device support

Add code and configuration to enable new USB device support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

desowin commented on include/zephyr/usb/usbd.h:
```diff
@@ -0,0 +1,717 @@
+		       uint8_t iface, uint8_t alternate);
+
+	/** USB control request handler to device */
+	int (*control_to_dev)(struct usbd_class_node *const node,
```
Separating the to device and to host might seem good, but does not really comply to Universal Serial Bus Specification Revision 2.0, 9.3.1 bmRequestType
> This bitmapped field identifies the characteristics of the specific request. In particular, this field identifies the direction of data transfer in the second phase of the control transfer. The state of the Direction bit is ignored if the wLength field is zero, signifying there is no Data stage.

The device has to ignore the Direction bit when wLength is 0. Having separate handlers effectively prohibit the device from ignoring Direction bit.

desowin commented on subsys/usb/device_next/usbd_ch9.c:
```diff
@@ -0,0 +1,871 @@
+		return 0;
+	}
+
+	ctrl_set_req_compl(uds_ctx, sreq_set_address_cb, NULL);
```
Having a generic callback mechanism for the **ONLY** request that works differently seems like an overkill to me. The `SET ADDRESS` should be handled separately as stated in Universal Serial Bus Specification Revision 2.0, 9.4.6 Set Address
> The USB device does not change its device address until after the Status stage of this request is completed successfully. Note that this is a difference between this request and all other requests. For all other requests, the operation indicated must be completed before the Status stage.


The reason why status stage is there is to allow host know that whole processing is done. The device firmware should not even attempt to get a notification that the status stage is finished as it is completely unnecessary (the **ONLY** exception is `SET ADDRESS`) from device perspective. When the device allows status stage to proceed, it means that the processing from device perspective is over and no further action will be taken regarding the request.

Note that if there is some long action after the control transfer, like for example flashing after `DFU DNLD`, it is up to the class to provide polling mechanism for host to determine when the action has finished (DFU does so with `GETSTATUS`).

desowin commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1112 @@
+		 * Next state depends on the direction and
+		 * the value of wLength.
+		 */
+		if (length && udc_data_stage_to_device(buf)) {
```
Universal Serial Bus Specification Revision 2.0 9.3.1
> The state of the Direction bit is ignored if the wLength field is zero, signifying there is no Data stage.

The simplest way to comply to the specification is to change the if chain to start with `if (length == 0)` then `else if (udc_data_stage_to_device(buf))` and finally `else` (data stage from device).

desowin reviewed (changes requested):
I think having the status stage completion handler is overkill because it can only **ever** be used by `SET ADDRESS`.

desowin commented on drivers/usb/udc/udc_nrf.c:
```diff
@@ -0,0 +1,817 @@
+		udc_submit_event(udc_nrf_dev, UDC_EVT_RESET, 0, NULL);
+		break;
+	case NRFX_USBD_EVT_SOF:
+		LOG_DBG("SoF");
```
Enabling debug level tracing essentially pollutes the log with SoF (every 1ms) and the logging becomes unusable. Please remove this line.

desowin reviewed (changes requested):
I cannot get the CDC ACM to receive data correctly. The first DATA packet on OUT endpoint is printed on console, while the second DATA packet is ACKed on the bus but not printed on console. Subsequent DATA packets are NAKed.

Enabling debug logging leads to MPU Fault Data Access Violation, the fault output is not really helpful (however, it mentions `Stacking error (context area might be not valid)`, so that's probably all the fault handler can do).

desowin commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -0,0 +1,1066 @@
+#define CDC_ACM_RX_FIFO_BUSY		4
+#define CDC_ACM_TX_FIFO_BUSY		5
+
+#define CONFIG_USBD_CDC_ACM_STACK_SIZE	512
```
This stack is too low for default configuration builds. The "only first packet works" issues are gone when I increase the stack to 1024.

The build command I use is:
`OVERLAY_CONFIG=usbd_next_prj.conf west build --pristine -b nrf52840dk_nrf52840 -d build-cdc-acm-next ~/zephyrproject/zephyr/samples/subsys/usb/cdc_acm`

desowin commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -0,0 +1,1066 @@
+	uint32_t done;
+
+	if (!check_wq_ctx(dev)) {
+		LOG_WRN("Invoked by inappropriate context");
```
If the code is not supposed to be called from other context at all, then check should really be an __ASSERT() (because it denotes programmer bug).

jfischer-no replied:
ASSERT are disable per default and do not help either.

desowin replied:
This is why you develop with ASSERT enabled and only put things without side effects within the assert. Generally, it is nice to have asserts available in the code **before** spotting a bug, as first thing after a bug is found, is to just build with asserts and check if it helps.

See https://stackoverflow.com/a/1245881/2200062 for more information. I believe an ASSERT here makes perfect sense and using `LOG_WRN` is actually an anti-pattern.

jfischer-no replied:
Also, generally, it is nice to have logging available in the code per default, one do not need to turn anything on and will see it immediately.

jfischer-no committed:
samples: usb: console: enable optional new USB device support

Add code and configuration to enable new USB device support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc_nrf: do not abuse HAL events for the driver's concerns.

HAL events were also used for the shim driver's concerns
during the prototyping of API. Fix it now and use specific
shim driver events. That also allows new transfers to be
triggered from a single point.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

desowin commented on subsys/usb/device_next/usbd_endpoint.c:
```diff
@@ -0,0 +1,195 @@
+
+	usbd_ep_bm_clear(ep_bm, ep);
+
+	ret = udc_ep_dequeue(dev, ep);
```
This should be called only if the EP was enqueued. Unconditionally dequeueing the EP leads to `<err> udc_nrf: ep 0x01 queue is empty` after disconnecting the device running CDC ACM sample.

desowin replied:
If the check is supposed to happen in the driver, then the `udc_ep_dequeue()` API documentation should mention it. Otherwise, there'll be some drivers that do handle it and some that don't.

carlescufi commented on drivers/usb/udc/Kconfig:
```diff
@@ -0,0 +1,52 @@
+
+
+menuconfig UDC_DRIVER
+	bool "USB device controller driver (NEW UDC API)"
```
`select EXPERIMENTAL`

carlescufi replied:
```suggestion
	bool "USB device controller driver [EXPERIMENTAL]"
```

jfischer-no commented on samples/subsys/usb/cdc_acm/usbd_next_prj.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_USB_DEVICE_STACK_NEXT=y
```
It is not a configuration overlay, `cmake -DBOARD=frdm_k64f -DCONF_FILE=usbd_next_prj.conf ../` should work fine.

jfischer-no commented on drivers/usb/udc/udc_common.c:
```diff
@@ -0,0 +1,1112 @@
+		goto device_speed_error;
+	}
+
+	if (api->device_speed) {
```
Because FS-only supports only FS, so it can not change "during connection" to the host.

desowin replied:
The fact that it can be `NULL` for FS devices should be mentioned in the API documentation.

jfischer-no committed:
drivers: udc_nrf: implement method to distiguish hal and shim events

Implement method to distiguish hal driver and UDC shim driver events.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -0,0 +1,1125 @@
+	struct cdc_acm_uart_data *data = dev->data;
+
+	atomic_set_bit(&data->state, CDC_ACM_CLASS_ENABLED);
+	LOG_WRN("Configuration enabled");
```
Yes, have forgotten it again...

desowin commented on drivers/usb/udc/udc_nrf.c:
```diff
@@ -140,6 +161,7 @@ static void udc_event_xfer_in(const struct device *dev,
 		buf = udc_buf_get(dev, ep, true);
 		if (buf == NULL) {
 			LOG_ERR("ep 0x%02x queue is empty", ep);
+			__ASSERT(false, "ep 0x%02x queue is empty", ep);
```
Asserting on false is not that useful. It would be better to make it `__ASSERT(buf != NULL, "ep 0x%02x queue is empty", ep);`. Also I think it would be better to place assert before `if (buf == NULL)`

desowin commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -0,0 +1,1131 @@
+
+	if (!check_wq_ctx(dev)) {
+		LOG_WRN("Invoked by inappropriate context");
+		__ASSERT(true, "Invoked by inappropriate context");
```
Asserting on true makes no sense. Condition `true` is always true and therefore always is fulfilled.

The assert should be `__ASSERT(check_wq_ctx(dev), "Invoked by inappropriate context");` and be placed before the if statement.

jfischer-no replied:
No, it should be `__ASSERT(false, "Invoked by inappropriate context");`, alternative is no ASSERT at all.

desowin reviewed (approved):
Time to merge this, despite some yet unresolved issues. This is experimental after all and will stay experimental for a while.
